test/dynamo/test_export.py

```diff
@@ -1794,6 +1799,40 @@ def forward(self, x):
         dynamo_result = out_graph(inp)
         self.assertEqual(dynamo_result, m(inp))
 
+    @config.patch(dynamic_shapes=True)
+    def test_export_raise_guard_full_constraint(self):
+        y = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            if x.shape[0] == 3:
+                return x.sin()
+            return x.cos()
+
+        torch._dynamo.export(my_dyn_fn, y)
+        torch._dynamo.mark_dynamic(y, 0)
+
+        with self.assertRaises(
+            torch._dynamo.exc.InternalTorchDynamoError,
+        ):
+            torch._dynamo.export(my_dyn_fn, y)
+
+    @config.patch(dynamic_shapes=True)
+    def test_export_raise_guard_partial_constraint(self):
+        y = torch.randn([3, 3, 3])
+
+        def my_dyn_fn(x):
+            if x.shape[0] > 3:
+                return x.sin()
+            return x.cos()
+
+        torch._dynamo.export(my_dyn_fn, y)
+        torch._dynamo.mark_dynamic(y, 0)
+
+        with self.assertRaises(
+            torch._dynamo.exc.InternalTorchDynamoError,
+        ):
+            torch._dynamo.export(my_dyn_fn, y)
```

Comment:
Some more stuff to test:

- Relational guards
- Multiple dynamic dims
- 0/1 handling

Reply:
Added most of the above - what do you want to see tested in 0/1?

Reply:
This is not so much a "test" as it is "documenting the behavior". If I mark a dimension as dynamic, do I assume it is non-zero/non-one? I think it is best if we do NOT make this negative inference. This is more code though.

Reply:
You can test it by just having a condition on the dynamic dim that it is != 1, and having this appropriately throw?